hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMRecoverLeaseRequest.java

```diff
@@ -186,60 +179,56 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
-      throws IOException {
-
+  private RecoverLeaseResponse doWork(OzoneManager ozoneManager,
+      long transactionLogIndex) throws IOException {
     final long volumeId = omMetadataManager.getVolumeId(volumeName);
-    final long bucketId = omMetadataManager.getBucketId(
-        volumeName, bucketName);
+    final long bucketId = omMetadataManager.getBucketId(volumeName, bucketName);
     Iterator<Path> pathComponents = Paths.get(keyName).iterator();
     long parentID = OMFileRequest.getParentID(volumeId, bucketId,
         pathComponents, keyName, omMetadataManager,
         "Cannot recover file : " + keyName
             + " as parent directory doesn't exist");
     String fileName = OzoneFSUtils.getFileName(keyName);
-    dbFileKey = omMetadataManager.getOzonePathKey(volumeId, bucketId,
-        parentID, fileName);
+    dbFileKey = omMetadataManager.getOzonePathKey(volumeId, bucketId, parentID, fileName);
 
     keyInfo = getKey(dbFileKey);
     if (keyInfo == null) {
-      throw new OMException("Key:" + keyName + " not found", KEY_NOT_FOUND);
+      throw new OMException("Key:" + keyName + " not found in keyTable", KEY_NOT_FOUND);
     }
-    final String clientId = keyInfo.getMetadata().remove(
-        OzoneConsts.HSYNC_CLIENT_ID);
-    if (clientId == null) {
+
+    final String writerId = keyInfo.getMetadata().get(OzoneConsts.HSYNC_CLIENT_ID);
+    if (writerId == null) {
       // if file is closed, do nothing and return right away.
-      LOG.warn("Key:" + keyName + " is already closed");
-      return null;
+      throw new OMException("Key: " + keyName + " is already closed", KEY_ALREADY_CLOSED);
     }
-    String openFileDBKey = omMetadataManager.getOpenFileName(
-            volumeId, bucketId, parentID, fileName, Long.parseLong(clientId));
-    if (openFileDBKey != null) {
-      commitKey(dbFileKey, keyInfo, fileName, ozoneManager,
-          transactionLogIndex);
-      removeOpenKey(openFileDBKey, fileName, transactionLogIndex);
+
+    dbOpenFileKey = omMetadataManager.getOpenFileName(
+        volumeId, bucketId, parentID, fileName, Long.parseLong(writerId));
+    openKeyInfo = omMetadataManager.getOpenKeyTable(getBucketLayout()).get(dbOpenFileKey);
+    if (openKeyInfo == null) {
+      throw new OMException("Open Key " + dbOpenFileKey + " not found in openKeyTable", KEY_NOT_FOUND);
+    }
+
+    if (openKeyInfo.getMetadata().containsKey(OzoneConsts.LEASE_RECOVERY)) {
+      LOG.debug("Key: " + keyName + " is already under recovery");
+    } else {
+      openKeyInfo.getMetadata().put(OzoneConsts.LEASE_RECOVERY, "true");
+      openKeyInfo.setUpdateID(transactionLogIndex, ozoneManager.isRatisEnabled());
+      openKeyInfo.setModificationTime(Time.now());
+      // Add to cache.
+      omMetadataManager.getOpenKeyTable(getBucketLayout()).addCacheEntry(
+          new CacheKey<>(dbOpenFileKey), CacheValue.get(transactionLogIndex, openKeyInfo));
```

Comment:
nit: we could use the cleaner form instead now that @szetszwo added it in HDDS-7969:

https://github.com/apache/ozone/commit/bd900ae9db8abf819b28a29551cff87259700afc#diff-ac3f28f30c26932cfbaf4270f5f0862b973b4e3e5b9b11ae78e725c306736fcdL324-L326

Reply:
i.e.

```suggestion
      // Add to cache.
            omMetadataManager.getOpenKeyTable(getBucketLayout()).addCacheEntry(
          dbOpenFileKey, openKeyInfo, transactionLogIndex);
```